model-optimizer/mo/front_ng/pipeline.py

```diff
@@ -0,0 +1,113 @@
+        :param places_original: List[Place] Original model places
+        :param places_new: List[Place] New list of places
+        :return: True if new list of places is same as original
+        """
+        eq = len(places_original) == len(places_new)
+        if eq:
+            for item in places_original:
+                found = [x for x in places_new if x['node'].is_equal(item)]
+                if not found:
+                    eq = False
+                    break
+        return eq
+
+    inputs_equal = True
+    if user_shapes:
+        inputs_equal = check_places(input_model.get_inputs(), user_shapes)
+
+    outputs_equal = True
+    if outputs:
+        outputs_equal = check_places(input_model.get_outputs(), outputs)
+    log.debug('Inputs are same: {}, outputs are same: {}'.format(
+        inputs_equal, outputs_equal))
+
+    if not inputs_equal and not outputs_equal:
+        # Use ExtractSubgraph
+        new_input_places = [x['node'] for x in user_shapes]
+        new_output_places = [x['node'] for x in outputs]
+        log.debug('Using extract subgraph')
+        input_model.extract_subgraph(new_input_places, new_output_places)
+    elif not inputs_equal:
+        new_input_places = [x['node'] for x in user_shapes]
+        log.debug('Using override_all_inputs')
+        input_model.override_all_inputs(new_input_places)
+    elif not outputs_equal:
+        new_output_places = [x['node'] for x in outputs]
+        log.debug('Using override_all_outputs')
+        input_model.override_all_outputs(new_output_places)
+
+    if user_shapes:
+        for user_shape in user_shapes:
+            if user_shape.get('shape') is not None:
+                input_model.set_partial_shape(
+                    user_shape['node'], PartialShape(user_shape['shape']))
+            if user_shape.get('data_type') is not None:
+                data_type = get_element_type(user_shape['data_type'])
+                log.debug('Set data type: {}'.format(data_type))
+                input_model.set_element_type(user_shape['node'], data_type)
+
+    # Set batch size
+    if argv.batch is not None and argv.batch > 0:
+        log.debug('Setting batch size to {}'.format(argv.batch))
+        for place in input_model.get_inputs():
+            old_partial_shape = input_model.get_partial_shape(place)
+            new_shape = []
+            old_shape_converted = []
+            joined_name = ' '.join(place.get_names())
+            if old_partial_shape.rank.is_static:
+                for i in range(old_partial_shape.rank.get_length()):
+                    # Assume batch size is always 1-st dimension in shape
+                    # Keep other dimensions unchanged
+                    new_shape.append(Dimension(argv.batch) if i == 0 else
+                                     old_partial_shape.get_dimension(i))
+                    old_shape_converted.append(
+                        old_partial_shape.get_dimension(i))
```

Comment:
let's create a new function to convert a PartialShape to the list or numpy array.  These loops looks awkward.

So you first convert the `old_partial_shape` to numpy array, update the first dimension and then run the `validate_batch_in_shape`. This way you don't need the `for` loop from line 93

Reply:
Done